File: app/src/main/java/com/example/fondodepantalla/FragmentosAdministrador/InicioAdmin.java
package com.example.fondodepantalla.FragmentosAdministrador;

import android.graphics.Color;
import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.fondodepantalla.R;
import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.components.MarkerView;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.formatter.ValueFormatter;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.utils.MPPointF;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.ArrayList;
import java.util.List;

public class InicioAdmin extends Fragment {

    BarChart barChartTareas, barChartInventario;
    FirebaseFirestore db;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_inicio_admin, container, false);

        barChartTareas = view.findViewById(R.id.barChart);
        barChartInventario = view.findViewById(R.id.barChartInventario);

        db = FirebaseFirestore.getInstance();

        cargarDatosTareas();
        cargarDatosInventario();

        return view;
    }

    // ------------------- TAREAS -------------------
    private void cargarDatosTareas() {
        db.collection("tareas")
                .get()
                .addOnSuccessListener(querySnapshot -> {
                    int completadas = 0;
                    int enProceso = 0;
                    int faltantes = 0;

                    for (DocumentSnapshot doc : querySnapshot) {
                        Boolean taken = doc.getBoolean("taken");
                        Long progress = doc.getLong("progress");
                        boolean evidenciaExiste = doc.contains("evidencias");

                        if (taken != null && taken) {
                            if (progress != null && progress == 4 && evidenciaExiste) {
                                completadas++;
                            } else {
                                enProceso++;
                            }
                        } else {
                            faltantes++;
                        }
                    }

                    mostrarGraficoTareas(completadas, enProceso, faltantes);
                });
    }

    private void mostrarGraficoTareas(int completadas, int enProceso, int faltantes) {
        List<BarEntry> entries = new ArrayList<>();
        entries.add(new BarEntry(0f, completadas));
        entries.add(new BarEntry(1f, enProceso));
        entries.add(new BarEntry(2f, faltantes));

        BarDataSet set = new BarDataSet(entries, "Tareas");
        // Colores pastel
        set.setColors(Color.rgb(186, 255, 201), Color.rgb(255, 255, 186), Color.rgb(255, 186, 186));
        set.setValueTextColor(Color.BLACK);
        set.setValueTextSize(14f);

        BarData data = new BarData(set);
        data.setBarWidth(0.9f);

        barChartTareas.setData(data);
        barChartTareas.getDescription().setEnabled(false);

        barChartTareas.getXAxis().setDrawLabels(true);
        barChartTareas.getXAxis().setGranularity(1f);
        barChartTareas.getXAxis().setValueFormatter(new ValueFormatter() {
            @Override
            public String getFormattedValue(float value) {
                int index = (int) value;
                switch (index) {
                    case 0: return "Completadas";
                    case 1: return "En proceso";
                    case 2: return "Faltantes";
                    default: return "";
                }
            }
        });

        barChartTareas.setFitBars(true);
        barChartTareas.invalidate();
    }

    // ------------------- INVENTARIO -------------------
    private void cargarDatosInventario() {
        db.collection("inventario")
                .get()
                .addOnSuccessListener(querySnapshot -> {
                    List<BarEntry> entries = new ArrayList<>();
                    List<String> nombres = new ArrayList<>();
                    int index = 0;

                    for (DocumentSnapshot doc : querySnapshot) {
                        Long cantidad = doc.getLong("cantidad");
                        String nombre = doc.getString("nombre");
                        if (cantidad != null && nombre != null) {
                            entries.add(new BarEntry(index, cantidad));
                            nombres.add(nombre);
                            index++;
                        }
                    }

                    mostrarGraficoInventario(entries, nombres);
                });
    }

    private void mostrarGraficoInventario(List<BarEntry> entries, List<String> nombres) {
        BarDataSet set = new BarDataSet(entries, "Inventario");

        // Colores pastel variados
        int[] coloresPastel = {
                Color.rgb(255, 179, 186),
                Color.rgb(255, 223, 186),
                Color.rgb(255, 255, 186),
                Color.rgb(186, 255, 201),
                Color.rgb(186, 225, 255),
                Color.rgb(201, 186, 255)
        };
        List<Integer> listaColores = new ArrayList<>();
        for (int i = 0; i < entries.size(); i++) {
            listaColores.add(coloresPastel[i % coloresPastel.length]);
        }
        set.setColors(listaColores);

        // No mostrar valores arriba de la barra
        set.setDrawValues(false);

        BarData data = new BarData(set);
        data.setBarWidth(0.9f);

        barChartInventario.setData(data);
        barChartInventario.getDescription().setEnabled(false);
        barChartInventario.getXAxis().setDrawLabels(false);
        barChartInventario.setFitBars(true);

        // MarkerView para mostrar nombre y cantidad al tocar
        barChartInventario.setMarker(new MarkerView(requireContext(), R.layout.marker_view) {
            @Override
            public void refreshContent(com.github.mikephil.charting.data.Entry e, Highlight highlight) {
                int i = (int) e.getX();
                String texto = nombres.get(i) + ": " + (int)e.getY();
                ((android.widget.TextView)findViewById(R.id.tvMarker)).setText(texto);
                super.refreshContent(e, highlight);
            }

            @Override
            public MPPointF getOffset() {
                return new MPPointF(-(getWidth()/2f), -getHeight());
            }
        });

        barChartInventario.invalidate();
    }
}

File: app/src/main/java/com/example/fondodepantalla/FragmentosAdministrador/InicioAdmin.kt
package com.example.fondodepantalla.FragmentosAdministrador

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.unit.dp
import androidx.fragment.app.Fragment
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll

class InicioAdmin : Fragment() {

    private val db = FirebaseFirestore.getInstance()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = ComposeView(requireContext()).apply {
        setContent {
            MaterialTheme {
                Surface(modifier = Modifier.fillMaxSize()) {
                    InicioAdminScreen(db)
                }
            }
        }
    }
}

@Composable
fun InicioAdminScreen(db: FirebaseFirestore) {
    var tareas by remember { mutableStateOf(listOf<Int>()) }
    var inventario by remember { mutableStateOf(listOf<Pair<String, Int>>()) }

    LaunchedEffect(Unit) {
        tareas = cargarTareas(db)
        inventario = cargarInventario(db)
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState()) //permite hacer scroll
            .padding(16.dp)
            .background(MaterialTheme.colorScheme.background),
        verticalArrangement = Arrangement.spacedBy(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text("Panel general", style = MaterialTheme.typography.titleLarge)

        if (tareas.isEmpty() && inventario.isEmpty()) {
            CircularProgressIndicator()
        } else {
            if (tareas.isNotEmpty()) {
                TarjetaGrafico(
                    titulo = "Progreso de tareas",
                    etiquetas = listOf("Completadas", "En proceso", "Faltantes"),
                    valores = tareas
                )
            }

            if (inventario.isNotEmpty()) {
                TarjetaGraficoBarras(
                    titulo = "Inventario general",
                    datos = inventario
                )
            }
        }
    }
}

// -------------------- FIRESTORE --------------------

suspend fun cargarTareas(db: FirebaseFirestore): List<Int> {
    val snapshot = db.collection("tareas").get().await()
    var completadas = 0
    var enProceso = 0
    var faltantes = 0

    for (doc in snapshot) {
        val taken = doc.getBoolean("taken") ?: false
        val progress = doc.getLong("progress") ?: 0
        val evidencia = doc.contains("evidencias")

        when {
            taken && progress == 4L && evidencia -> completadas++
            taken -> enProceso++
            else -> faltantes++
        }
    }
    return listOf(completadas, enProceso, faltantes)
}

suspend fun cargarInventario(db: FirebaseFirestore): List<Pair<String, Int>> {
    val snapshot = db.collection("inventario").get().await()
    return snapshot.mapNotNull {
        val nombre = it.getString("nombre")
        val cantidad = it.getLong("cantidad")?.toInt()
        if (nombre != null && cantidad != null) nombre to cantidad else null
    }
}

// -------------------- UI --------------------

@Composable
fun TarjetaGrafico(titulo: String, etiquetas: List<String>, valores: List<Int>) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(MaterialTheme.colorScheme.surface),
        shape = RoundedCornerShape(20.dp),
        elevation = CardDefaults.cardElevation(6.dp)
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Text(titulo, style = MaterialTheme.typography.titleMedium)
            Spacer(Modifier.height(4.dp))
            AnimatedBarChart(etiquetas, valores)
        }
    }
}

@Composable
fun TarjetaGraficoBarras(titulo: String, datos: List<Pair<String, Int>>) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(MaterialTheme.colorScheme.surface),
        shape = RoundedCornerShape(20.dp),
        elevation = CardDefaults.cardElevation(6.dp)
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Text(titulo, style = MaterialTheme.typography.titleMedium)
            AnimatedBarChart(datos.map { it.first }, datos.map { it.second })
        }
    }
}

@Composable
fun AnimatedBarChart(labels: List<String>, values: List<Int>) {
    val maxValue = values.maxOrNull()?.toFloat() ?: 1f
    val pastelColors = listOf(
        Color(0xFFBAFFC9), Color(0xFFFFFFBA), Color(0xFFFFBABA),
        Color(0xFFBADCFF), Color(0xFFD5BAFF)
    )

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        values.forEachIndexed { index, value ->
            val progress = (value / maxValue)
            val animatedProgress = animateFloatAsState(targetValue = progress).value

            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(
                    text = labels[index],
                    modifier = Modifier.width(110.dp),
                    style = MaterialTheme.typography.bodyMedium
                )

                Box(
                    modifier = Modifier
                        .weight(1f)
                        .height(22.dp)
                        .background(
                            color = pastelColors[index % pastelColors.size].copy(alpha = 0.3f),
                            shape = RoundedCornerShape(50)
                        )
                ) {
                    Box(
                        modifier = Modifier
                            .fillMaxHeight()
                            .fillMaxWidth(animatedProgress)
                            .background(
                                color = pastelColors[index % pastelColors.size],
                                shape = RoundedCornerShape(50)
                            )
                    )
                }
                Spacer(Modifier.width(8.dp))
                Text("${value}")
            }
        }
    }
}
